algebra/direct_limit.lean

```diff
@@ -0,0 +1,290 @@
+import algebra.ring order.lattice
+
+universes u v
+
+open lattice
+
+instance is_ring_hom.to_is_add_group_hom {α : Type*} {β : Type*} [ring α] [ring β]
+  (f : α → β) [is_ring_hom f] : is_add_group_hom f :=
+⟨λ _ _, is_ring_hom.map_add f⟩
+
+class directed_order (α : Type u) extends has_sup α, partial_order α :=
+(le_sup_left : ∀ a b : α, a ≤ a ⊔ b)
+(le_sup_right : ∀ a b : α, b ≤ a ⊔ b)
+
+theorem le_sup_left {α : Type u} [directed_order α] {x y : α} : x ≤ x ⊔ y :=
+directed_order.le_sup_left x y
+
+theorem le_sup_right {α : Type u} [directed_order α] {x y : α} : y ≤ x ⊔ y :=
+directed_order.le_sup_right x y
+
+variables {ι : out_param (Type u)} [inhabited ι] [directed_order ι]
+variables {G : ι → Type v}
+
+class directed_system (f : Π i j, i ≤ j → G i → G j) : Prop :=
+(Hcomp : ∀ i j k hij hjk x, f j k hjk (f i j hij x) = f i k (le_trans hij hjk) x)
```

Comment:
This is a non-standard definition of a directed system. You have omitted the axiom that if `i = j` then the induced map `G i -> G j` is the identity. Here's a sketch of why dropping the axiom doesn't affect things much -- the `Hcomp` axiom implies that `e_i : G i -> G i` is idempotent (that is, equals its own square) and commutes with all other maps to and from `G i`. In particular, one can replace `G i` with `G' i`, the image of `e i`, and get a modified system which now does satisfy the standard axioms for a direct system, and whose direct limit is the limit computed in the file here. In particular, dropping the identity axiom amounts to nothing more than allowing some extra noise. The problem I have with the definition as it stands is that it is non-standard -- this is not what a mathematician calls a direct system so I would strongly argue that it should not be what mathlib calls a direct system -- the axiom should be added. It is never used in the proofs here but it is used in explicit computations of direct limits and it will lead to confusion if it is not added. If it is decided not to add the axiom then I would strongly recommend changing the name of the function, because what we have here is not a direct system, even though it shares many of the properties of a direct system.

Reply:
I've realised another way of formalising what I'm saying. There are theorems about direct limit (for example -- if the partially ordered set is discrete, so i <= j iff i = j, then the direct limit is the disjoint union) which will not be provable in Lean if we use this non-standard definition.

Reply:
@kbuzzard I don't know what you're talking about; a discrete set isn't directed.

Reply:
Aah yes you're right. I'd better stick with the first way :-)